tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,767 @@
+        SigningKeyConfig::kms { config, .. } => Ok(config
+            .available_keys
+            .as_ref()
+            .context(error::KeyCreation)?
+            .clone()),
+        SigningKeyConfig::ssm { .. } => Ok(HashMap::new()),
+    }
+}
+
+/// Creates keys using data stored in SigningKeyConfig enum
+/// Output: This method will has an in-place output meaning it edits key-id
+/// If key-id is populated, it will not change. Otherwise, it will be populated with a key-id of an available key
+async fn create_keys(signing_key_config: &mut SigningKeyConfig) -> Result<()> {
+    match signing_key_config {
+        SigningKeyConfig::file { path } => (),
+        SigningKeyConfig::kms { key_id, config, .. } => {
+            // QUESTION: keyid must remain at top level (we could encapsulate in config, but it would alter keyid reading for pubsys (which we alredy are technically doing by changing it to Option))
+            // TODO: don't override if key_id is not none 
+            let keyid_raw = config.create_kms_keys().await?;
+            *key_id = Some(keyid_raw);
+        } // QUESTION: unfortunately setting keyid for root key as well, but we don't use it
+        SigningKeyConfig::ssm { parameter } => (),
+    }
+    Ok(())
+}
+
+/// Must create a trait because can't directly implement a method for an struct in an external crate like KMSKeyConfig (which lives in pubsys-config/lib.rs)
+#[async_trait]
+trait KMSKeyConfigEx {
+    async fn create_kms_keys(&mut self) -> Result<String>;
+}
+
+/// Creates new KMS keys using cloudformation in regions specified
+/// Input Conditions: Alias+Region or AvailableKeys must be specified
+/// Output: Returns keyid and populates KMSKeyConfig with information about resources created
+/// available-keys starts as a map of pre-existing keyids:regions and will end as a map of pre-existing and generated keyids:regions
+/// key-stack-arns starts empty and will end as a map of keyids:stackarn if new keys are created
+#[async_trait]
+impl KMSKeyConfigEx for KMSKeyConfig {
+    async fn create_kms_keys(&mut self) -> Result<String> {
+        // Creating output params
+        let mut new_available_keys = HashMap::new();
+        let mut stack_arns = HashMap::new();
+
+        // TODO: if everything is null, do nothing (self.key_alias.is_none() && self.regions.is_some() && self.available_keys.is_none())
+        // Checking pre-conditions
+        if (self.key_alias.is_some() && self.regions.is_none())
+            || (self.key_alias.is_none() && self.regions.is_some())
+        {
+            error::KeyConfig.fail()?;
+        }
+        if self.available_keys.is_some() {
+            //available keys is either empty or populated with current values
+            new_available_keys = self
+                .available_keys
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: "available_keys",
+                })?
+                .clone();
+        }
```

Comment:
This should accomplish what you're trying to do:
```suggestion
        // Available keys list is either empty or populated with current values
        let mut new_available_keys = if let Some(available_keys) = self.available_keys {
             available_keys
        } else {
             HashMap::new();
        }
```
(you'd also no longer need the `new_available_keys` declaration from above)